### ISSUE TITLE
[Question][NVIDIA]: How to save memory in GPT pretraining? 

### ISSUE THREAD

**Issue by Wong4j** [PaddlePaddle/PaddleNLP#7254]
### 请提出你的问题

We are integrating TransformerEngine with PaddleNLP and need to verify fp8 training convergence for GPT models. Now we have already verify the convergence of 1.3B/5B/13B GPT model on 64xH100 GPUs.

The next models we want to verify are 89/175B GPT. But it always got OOM no matter how to setup the training. I have tried `FSDP stage3`, `Tensor Parallel` and recompute the full TransformerLayer. But they are all OOM w/ or w/o TransformerEngine on 128xH100 GPUs.

It's there any way to further reduce the momery usage in GPT training? 

GPU：128xH100 80GB
Config:
- BS=1 DP=1 TP=1 FSDP=128 stage3 recompute=full
- BS=1 DP=1 TP=8 FSDP=16 stage2 recompute=full
- BS=1 DP=2 TP=8 FSDP=8 stage2 recompute=full

Command example:
```
python -u  -m paddle.distributed.launch \
    --gpus "0,1,2,3,4,5,6,7" \
    --log_dir logs/89B/${log_dir} \
    --ips="${IP_STR}" \
    ${PNLP_PATH}/llm/gpt-3/run_pretrain.py \
    --model_type "gpt" \
    --model_name_or_path gpt3-89B-en \
    --tokenizer_name_or_path gpt3-89B-en \
    --input_dir "/workspace/llama_openwebtext" \
    --output_dir $OUT_DIR \
    --split 949,50,1 \
    --max_seq_length 1024 \
    --per_device_train_batch_size $BS \
    --per_device_eval_batch_size $BS \
    --tensor_parallel_degree $TP \
    --sharding_parallel_degree $FSDP \
    --sharding $STAGE \
    --pipeline_parallel_degree 1 \
    --fuse_attention_qkv 1 \
    --use_flash_attention 0 \
    --bf16  \
    --fp16_opt_level "O2"  \
    --scale_loss 1024 \
    --learning_rate 0.00001 \
    --min_learning_rate 0.000005 \
    --max_steps 40000 \
    --save_steps 5000 \
    --weight_decay 0.01 \
    --warmup_ratio 0.01 \
    --max_grad_norm 1.0 \
    --logging_steps 100 \
    --dataloader_num_workers 1 \
    --hidden_dropout_prob=0.1 \
    --attention_probs_dropout_prob=0.1 \
    --eval_steps 1000 \
    --report_to "visualdl" \
    --disable_tqdm true \
    --gradient_accumulation_steps 1 \
    --do_train \
    --do_eval \
    --continue_training 0 \
    --recompute 1 \
    --recompute_granularity full \
    --device "gpu"
```

**Comment by DrownFish19** [comment]
1. `sharding_parallel_degree $FSDP` indicates sharding parallel among `degree` devices, optimizer states / gradients / parameters will be paralleled for 'degree' partitions. If you prefer communication within one SERVER, `sharding_parallel_degree` can be set to 8.
5. `pipeline_parallel_degree` can parallel model parameters with a pipeline among devices, which can reduce memory consumption for each device.
6. `tensor_parallel_degree` can parallel model parameters with splitting `weight` parameters among device, which can reduce memory.
7. for all device, `sharing_parallel_degree * tensor_parallel_degree * pipeline_parallel_degree * data_parallel_degree = all devices`, in which the `data_parallel_degree` is computed according other parallel_degree configs.

**Comment by Wong4j** [comment]
感谢回复。目前测试发现128卡，设置TP=8 PP=16，可以跑起来GPT 89/175模型。但nccl通信是瓶颈，我这边之后会继续调查。

**Comment by Wong4j** [comment]
@DrownFish19 FSDP将参数/梯度/优化器状态切分，每个GPU只存一部分，目的就是为了省内存，请问为什么说不能减少内存消耗呢？
并且其他框架可以用pure FSDP stage3来跑175B模型，但我用PaddleNLP跑会OOM。

**Comment by DrownFish19** [comment]
> @DrownFish19 FSDP将参数/梯度/优化器状态切分，每个GPU只存一部分，目的就是为了省内存，请问为什么说不能减少内存消耗呢？ 并且其他框架可以用pure FSDP stage3来跑175B模型，但我用PaddleNLP跑会OOM。

1. 之前的表述存在问题，现在已经修改：`sharding_parallel_degree $FSDP` indicates sharding parallel among `degree` devices, optimizer states / gradients / parameters will be paralleled for 'degree' partitions. If you prefer communication within one SERVER, `sharding_parallel_degree` can be set to 8.
2. 请问后续有尝试stage3，sharding_parallel_degree=8 测试GPT3-89B吗？

**Comment by DrownFish19** [comment]
@Wong4j Paddle当前执行sharding stage3需要模型先初始化，当模型初始化显存占用超过最大显存时，会出现OOM。后续会针对这种情况进行处理。

**Comment by Wong4j** [comment]
@DrownFish19 
OK，实际测试确实stage3在13/89/175B都会OOM，目前在稍大的模型上都不可用。
128卡，TP8+PP16都能跑，但是目前观察到通信时间是bottleneck，占了绝大多数时间。这样的话FP8训练就意义不大，因为FP8只用在GEMM计算上，通信仍然是BF16。所以后续我需要调查PP的通信的问题，请问你们有经验吗？
![image](https://github.com/PaddlePaddle/PaddleNLP/assets/21985950/191fbd22-6d6c-4744-af86-e9e501b62865)

**Comment by Wong4j** [comment]
设置较大的gradient accumulation steps可以明显提升性能，先关掉issue。
